File: binary_tree_inorder_traversal_iterative.cpp
#include<stdlib.h>
#include<stdio.h>
#include<stack>

using namespace std;

struct node{
    int val;
    struct node* left;
    struct node* right;
};


void in_order(struct node* root){
    stack<struct node*> s;
    s.push(root);
    while(!s.empty()){
        struct node* t = s.top();
        if(t->left){
            s.push(t->left);
            continue;
        }
        printf("%d ", t->val);
        while(!s.empty() && !s.top()->right){
            printf("%d ", s.top()->val);
            s.pop();
        }
        if(s.top()->right){
            printf("%d ", s.top()->val);
            t=s.top()->right;
            s.pop();
            s.push(t);
        }
    }
    printf("\n");
}


int main(){
    struct node one={.val=1,.left=NULL,.right=NULL};
    struct node two={.val=2,.left=NULL,.right=NULL};
    struct node three={.val=3,.left=NULL,.right=NULL};
    one.right=&two;
    two.left=&three;
    in_order(&one);
    return 0;
}

File: master/binary_tree_inorder_traversal_iterative.py

class Node(object):
    def __init__(self,val):
        self.val = val
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def inorderTraversal(self, root):
        res = []
        if root is None:
            return res
        stack = [root]
        while len(stack)>0:
            while stack[-1].left is not None:
                stack.append(stack[-1].left)
            while len(stack)>0 and stack[-1].right is None:
                res.append(stack[-1].val)
                stack.pop()
            if not len(stack)==0:
                res.append(stack[-1].val)
                n = stack.pop()
                stack.append(n.right)
        return res


if __name__ == "__main__":
    n1 = Node(1)
    n2 = Node(2)
    n3 = Node(3)
    n1.right = n2
    n2.left = n3
    print Solution().inorderTraversal(n1)
